cmd/libsnap-confine-private/device-cgroup-support.c

```diff
@@ -702,15 +694,15 @@ static int sc_udev_open_cgroup_v1(const char *security_tag, int flags, sc_cgroup
     const char *security_tag_relpath = security_tag;
     if (!from_existing) {
         /* Open snap.$SNAP_NAME.$APP_NAME relative to /sys/fs/cgroup/devices,
-         * creating the directory if necessary. Note that we always chown the
-         * resulting directory to root:root. */
-        sc_identity old = sc_set_effective_identity(sc_root_group_identity());
-        if (mkdirat(devices_fd, security_tag_relpath, 0755) < 0) {
-            if (errno != EEXIST) {
-                die("cannot create directory %s/%s/%s", cgroup_path, devices_relpath, security_tag_relpath);
+         * creating the directory if necessary. */
+        if (mkdirat(devices_fd, security_tag_relpath, 0755) == 0) {
+            /* the new directory must be owned by root:root. */
+            if (fchownat(devices_fd, security_tag_relpath, 0, 0, 0) < 0) {
```

Comment:
If a non-root process has privileges to write into this directory, a non-root process can probably race the `mkdirat` and `fchownat` calls to allow creating a directory or setting permissions through a symbolic link.

Reply:
I'm not sure I understand your point. I see that if a process could remove the `security_tag_relpath` we just created and recreate if as symbolic link before our call to `fchownat()`, then we'd be setting the root:root ownership on a different file. I will add `AT_SYMLINK_NOFOLLOW` to fix this.

Please let me know if this is enough, or if (as I suspect) I'm missing something from your remark :-)

Reply:
So, I was wrong on the `mkdirat` -- I forgot that it will return `EEXIST` rather than write through a last-component symlink.

`AT_SYMLINK_NOFOLLOW` for the `fchownat` will address that nicely, so long as the parent directories can't be controlled by untrusted processes. Thanks

Reply:
Good idea, I've added the flag!